### PR TITLE
[Model] Consolidate pooler implementations

### PR DESCRIPTION
**Pull request by DarkLight1337**
## Essential Elements of an Effective PR Description Checklist
- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.

## Purpose

Use `ClassifierPooler` in `as_seq_cls_model` and `JambaForSequenceClassification`. cc @noooop @maxdebayser 

cc @yecohn @tomeras91 can you help check whether the output of Jamba-reward remains the same? There are no tests that check the correctness of this model.

### CODE REVIEW

**Comment by DarkLight1337** [comment]
Let me fix some bugs locally first

**Comment by noooop** [comment]
I'm going to implement this todo next

```
# TODO: The reward weights themselves have float32 accuracy data, we
# would like to load them in fp32 to get that extra precision.
```

Related to #19925

**Review by maxdebayser — APPROVED** [review]
Nice refactoring!

**Comment by DarkLight1337** [comment]
Had to do another round of refactoring, PTAL again

File: vllm/model_executor/layers/pooler.py [hunk 1]
```diff
@@ -349,23 +355,159 @@ def from_config_with_defaults(
...
+        hidden_states: Union[torch.Tensor, list[torch.Tensor]],
+        pooling_metadata: PoolingMetadata,
+    ) -> PoolerOutput:
+        pooled_data = self.method.extract_states(hidden_states,
+                                                 pooling_metadata)
+        pooled_data = self.head(pooled_data, pooling_metadata)
+        return build_output(pooled_data)
+
+
+class StepPooler(BasePooler):
```
Comment (DarkLight1337) [inline vllm/model_executor/layers/pooler.py:484]: This doesn't really fit in with the other pooling types so I made this its own class

File: vllm/model_executor/layers/pooler.py [hunk 2]
```diff
@@ -31,165 +31,198 @@ class PoolingType(IntEnum):
...
+            if pooler_config.pooling_type is not None else pooling_type,
+            normalize=pooler_config.normalize
+            if pooler_config.normalize is not None else normalize,
+            softmax=pooler_config.softmax
+            if pooler_config.softmax is not None else softmax,
+            step_tag_id=pooler_config.step_tag_id
+            if pooler_config.step_tag_id is not None else step_tag_id,
+            returned_token_ids=pooler_config.returned_token_ids
+            if pooler_config.returned_token_ids is not None else
+            returned_token_ids,
```
Comment (Isotr0py) [inline vllm/model_executor/layers/pooler.py:63]:
```suggestion
            normalize=pooler_config.normalize or normalize,
            softmax=pooler_config.softmax or softmax,
            step_tag_id=pooler_config.step_tag_id or step_tag_id,
            returned_token_ids=pooler_config.returned_token_ids or returned_token_ids,
```
I think we can reduce the `if ... else ...` statement here.

**Review by Isotr0py — APPROVED** [review]
Overall LGTM!

**Comment by noooop** [comment]
Nice refactoring! +1 
This implementation can simplify a lot of code

File: vllm/model_executor/layers/pooler.py (hunk 2 above)
Comment (DarkLight1337) [reply vllm/model_executor/layers/pooler.py:63]: For `normalize` and `softmax`, it is important to distinguish between `None` (fall back to the default value) and `False` (don't use the default)

File: vllm/transformers_utils/config.py [hunk 3]
```diff
@@ -775,28 +773,6 @@ def try_get_generation_config(
             return None
 
 
-def get_classification_activation_function(config: PretrainedConfig):
```
Comment (DarkLight1337) [inline vllm/transformers_utils/config.py:778]: These functions aren't used anywhere outside of `pooler.py` so I moved them

File: vllm/model_executor/models/interfaces.py [hunk 4]
```diff
@@ -639,7 +639,7 @@ def supports_cross_encoding(
 def has_step_pooler(model: Union[type[object], object]) -> bool:
     """Check if the model uses step pooler."""
     return is_pooling_model(model) and any(
-        type(module).__name__ == "StepPool" for module in model.modules())
+        type(module).__name__ == "StepPooler" for module in model.modules())
```
Comment (DarkLight1337) [inline vllm/model_executor/models/interfaces.py:642]: This looks hacky. I'm planning to require models to define `pooler` as a `BasePooler` instance in the next PR so we can directly inspect `model.pooler` to get this information

**Review by maxdebayser — APPROVED** [review]
I like what you did there with forward_one and forward_all. In V0 were almost always doing the equivalent of forward_all. It's potentially more efficient as you can operate on the whole batch at once in the GPU. However it's also more difficult to generalize because not all pooling types are easy to implement on the full concatenated batch. Because of that, having in mind the hidden states processor feature, in the V1 embeddings PR I made the batch separation already at the model runner. The idea was that it would be possible to chain arbitrary HSPs and that the most flexible data representation would be the list of hidden states. But if the supported input type is made explicit with `forward_one` and `forward_all`, then we can delay the splitting of the batch later in the the pipeline.

**Comment by maxdebayser** [comment]
Maybe we could even do something like:

```
class PoolingMethod(nn.Module, ABC):

    @abstractmethod
    def forward_one(
        self,
        hidden_states: torch.Tensor,
        prompt_len: Optional[torch.Tensor] = None,
    ) -> torch.Tensor:
        """
        Note:
            `prompt_len=None` means `prompt_len=len(hidden_states)`.
        """
        raise NotImplementedError

class FullBatchPoolingMethod(PoolingMethod, ABC):

    @abstractmethod
    def forward_all(
        self,
        hidden_states: torch.Tensor,
        prompt_lens: torch.Tensor,
    ) -> Union[list[torch.Tensor], torch.Tensor]:
        raise NotImplementedError

```

**Comment by DarkLight1337** [comment]
Yeah I will further refine the Pooler interface in the next PR, see https://github.com/vllm-project/vllm/pull/20927#discussion_r2206644776

**Comment by DarkLight1337** [comment]
I'm also planning to include information about which APIs are supported in Pooler (`Pooler.get_pooling_params(task)` which returns `None` if the task is not supported) to support dynamic pooling config during inference, which removes the need for users to set `--task` at initialization time.

**Comment by DarkLight1337** [comment]
Are you able to help with debugging the accuracy issue?

**Comment by noooop** [comment]
> Are you able to help with debugging the accuracy issue?

Just looking at the code, it doesn’t seem like there should be an issue. Perhaps there’s a potential numerical precision problem? Or maybe this test case isn’t robust enough?

**Comment by DarkLight1337** [comment]
https://buildkite.com/vllm/ci/builds/24080/steps/canvas?jid=0198112d-331c-46d5-a405-fe58dd402871

File: vllm/model_executor/models/jamba.py [hunk 5]
```diff
@@ -564,29 +565,41 @@ class JambaForSequenceClassification(JambaForCausalLM):
...
-        self.score = nn.Linear(config.hidden_size, num_labels, bias=score_bias)
+
+        # TODO: The original reward weights have float32 accuracy data, we
+        # would like to load them in fp32 to get that extra precision.
+        # Currently weight_loader passes the weight which is already in bf16
+        self.score = nn.Linear(
+            config.hidden_size,
+            num_labels,
+            bias=score_bias,
+            dtype=torch.float32,
```
Comment (noooop) [inline vllm/model_executor/models/jamba.py:580]: try not set this dtype=torch.float32,

Comment (DarkLight1337) [inline vllm/model_executor/models/jamba.py:580]: The model uses `Qwen2ForCausalLM` architecture so it should not be related

**Comment by noooop** [comment]
set MTEB_RERANK_TOL = 0.01 

I will switch to a more robust test as soon as possible

**Comment by noooop** [comment]
We cannot rule out a potential numerical problem with this pr, but it would be very difficult to find it.

https://buildkite.com/vllm/ci/builds/24088/steps/canvas?sid=01981164-83cc-49e3-bc15-5b73f41a98ba
[2025-07-16T07:33:04Z] VLLM: torch.bfloat16 0.27263
[2025-07-16T07:33:04Z] SentenceTransformers: torch.float32 0.273
[2025-07-16T07:33:04Z] Difference: 0.00037000000000003697
[2025-07-16T07:33:05Z] PASSED

https://buildkite.com/vllm/ci/builds/23961/steps/canvas?sid=01980c3e-3f93-48c6-9c00-f25ec6b40f6a
[2025-07-15T06:42:55Z] VLLM: torch.bfloat16 0.27336
[2025-07-15T06:42:55Z] SentenceTransformers: torch.float32 0.273
[2025-07-15T06:42:55Z] Difference: -0.00035999999999997145
[2025-07-15T06:42:56Z] PASSED

this pr
[2025-07-16T06:24:12Z] VLLM: torch.bfloat16 0.26798
[2025-07-16T06:24:12Z] SentenceTransformers: torch.float32 0.273
[2025-07-16T06:24:12Z] Difference: 0.0050200000000000244
[2025-07-16T06:24:13Z] FAILED

**Comment by DarkLight1337** [comment]
Yeah that is what I mean, trying to find out why the score decreased in this PR

**Comment by noooop** [comment]
```
models/language/pooling/test_mxbai_rerank.py
- A:      -0.00035999999999997145
- B:       0.00037000000000003697
- C:       0.00019000000000002348
- this pr: 0.0050200000000000244


tomaarsen/Qwen3-Reranker-0.6B-seq-cls
- A:       -1.0000000000010001e-05
- B:       -0.0005000000000000004
- C:       -0.00028000000000000247
- this pr: -0.000490000000000046


models/language/pooling/test_qwen3_reranker.py
- A:       -0.00036000000000002697
- B:       -0.00017000000000000348
- C:       -0.0009700000000000264
- this pr: -0.000490000000000046


BAAI/bge-reranker-v2-gemma
- A:       0.0013299999999999979
- B:       0.0013299999999999979
- C:       0.0013299999999999979
- this pr: 0.0006399999999999739


cross-encoder/ms-marco-TinyBERT-L-2-v2
- A: -3.999999999998449e-05
- B: -3.999999999998449e-05
- C: -3.999999999998449e-05
- this pr: -3.999999999998449e-05

BAAI/bge-reranker-base
- A: 2.999999999997449e-05
- B: 2.999999999997449e-05
- C: 2.999999999997449e-05
- this pr: 2.999999999997449e-05

```

Maybe there is indeed a potential numerical stability problem, and it does not affect the bert-like model, 

BAAI/bge-reranker-v2-gemma is relatively stable and can be used for debugging


I think the difference is relatively small, we can first set MTEB_RERANK_TOL = 0.01 and then find it gradually .

**Comment by noooop** [comment]
There is a detail that may need pay attention. Because V1 does not support Gemma, V0 is used.

https://github.com/vllm-project/vllm/blob/d0dc4cfca48c2734da18ec42d6bba1346cbfc400/tests/models/language/pooling/test_bge_reranker_v2_gemma.py#L121-L124


So is the numerical of v1 more unstable?

**Comment by DarkLight1337** [comment]
I think I found that issue. On main branch, the data passed to the activation function is in float32, but in this PR it is in the models' dtype. Let's see if using float32 fixes the problem

**Comment by noooop** [comment]
[buildkite/ci/pr/language-models-test-extended-pooling](https://buildkite.com/vllm/ci/builds/24101#0198129f-9276-4121-a711-d1c59a3634f6) passed

BAAI/bge-reranker-v2-gemma
Difference: 0.0013299999999999979


The reranker test isn't completely useless
